receiver/httpdreceiver/scraper.go

```diff
@@ -0,0 +1,248 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpdreceiver
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/model/pdata"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/httpdreceiver/internal/metadata"
+)
+
+type httpdScraper struct {
+	logger     *zap.Logger
+	cfg        *Config
+	httpClient *http.Client
+}
+
+func newHttpdScraper(
+	logger *zap.Logger,
+	cfg *Config,
+) *httpdScraper {
+	return &httpdScraper{
+		logger: logger,
+		cfg:    cfg,
+	}
+}
+
+func (r *httpdScraper) start(_ context.Context, host component.Host) error {
+	httpClient, err := r.cfg.ToClient(host.GetExtensions())
+	if err != nil {
+		return err
+	}
+	r.httpClient = httpClient
+	return nil
+}
+
+func initMetric(ms pdata.MetricSlice, mi metadata.MetricIntf) pdata.Metric {
+	m := ms.AppendEmpty()
+	mi.Init(m)
+	return m
+}
+
+func addToIntMetric(metric pdata.NumberDataPointSlice, labels pdata.AttributeMap, value int64, ts pdata.Timestamp) {
+	dataPoint := metric.AppendEmpty()
+	dataPoint.SetTimestamp(ts)
+	dataPoint.SetIntVal(value)
+	if labels.Len() > 0 {
+		labels.CopyTo(dataPoint.Attributes())
+	}
+}
+
+func (r *httpdScraper) scrape(context.Context) (pdata.ResourceMetricsSlice, error) {
+	if r.httpClient == nil {
+		return pdata.ResourceMetricsSlice{}, errors.New("failed to connect to http client")
+	}
+
+	stats, err := r.GetStats()
+	if err != nil {
+		r.logger.Error("Failed to fetch httpd stats", zap.Error(err))
+		return pdata.ResourceMetricsSlice{}, err
+	}
+
+	rms := pdata.NewResourceMetricsSlice()
+	ilm := rms.AppendEmpty().InstrumentationLibraryMetrics().AppendEmpty()
+	ilm.InstrumentationLibrary().SetName("otel/httpd")
+	now := pdata.NewTimestampFromTime(time.Now())
+
+	uptime := initMetric(ilm.Metrics(), metadata.M.HttpdUptime).Sum().DataPoints()
+	connections := initMetric(ilm.Metrics(), metadata.M.HttpdCurrentConnections).Gauge().DataPoints()
+	workers := initMetric(ilm.Metrics(), metadata.M.HttpdWorkers).Gauge().DataPoints()
+	requests := initMetric(ilm.Metrics(), metadata.M.HttpdRequests).Sum().DataPoints()
+	traffic := initMetric(ilm.Metrics(), metadata.M.HttpdTraffic).Sum().DataPoints()
+	scoreboard := initMetric(ilm.Metrics(), metadata.M.HttpdScoreboard).Gauge().DataPoints()
+
+	u, err := url.Parse(r.cfg.Endpoint)
+	if err != nil {
+		r.logger.Error("Failed to find parse server name", zap.Error(err))
+		return pdata.ResourceMetricsSlice{}, err
+	}
+	serverName := u.Hostname()
```

Comment:
This is a very small amount of work, but logically it can be done just once (not each scrape). Should it be moved to `start()`?

Reply:
Good point. The `start()` method does not return an error though, so it makes handling the error case a bit awkward. 

I ended up attaching it to the config struct, because the value is extracted from the config. This also allows us to fail during config validation if necessary.